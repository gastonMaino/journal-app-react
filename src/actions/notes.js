import swal from 'sweetalert2';


import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
        dispatch(noteAdd(doc.id, newNote));
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const noteAdd = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        }

        const noteFirestore = { ...note };
        delete noteFirestore.id

        db.doc(`${uid}/journal/notes/${note.id}`).update(noteFirestore);

        dispatch(refreshNote(note.id, noteFirestore));

        swal.fire('Saved', note.title, 'success');
    }
}



export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active } = getState().notes

        swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                swal.showLoading();
            }
        })

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        swal.close();

        dispatch(startSaveNote(active))

    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));

        swal.fire('Completed', 'successfully note deleted', 'success')
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})