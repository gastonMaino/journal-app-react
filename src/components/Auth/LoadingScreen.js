import React from 'react'

export const LoadingScreen = () => {
    return (
        <div className='auth__register'>
            <div className='auth__register-preloader'></div>
            <p className='auth__register-text'>wait...</p>
        </div>
    )
}
