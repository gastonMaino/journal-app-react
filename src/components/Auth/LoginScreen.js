import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm/useForm'

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { ui } = useSelector(state => state);
    const { loading } = ui;


    const [values, handleInputChange] = useForm({
        email: '',
        password: ''
    })

    const { email, password } = values;

    const handleLogin = (e) => {
        e.preventDefault();

        dispatch(startLoginEmailPassword(email, password))
    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }


    return (
        <>
            <h1 className='auth__title mb-4'>Login</h1>

            <form
                onSubmit={handleLogin}
                className='animate__animated animate__fadeIn animate__fast'
            >
                <input
                    type='text'
                    placeholder='Enter your email'
                    name='email'
                    value={email}
                    className='auth__input'
                    onChange={handleInputChange}
                />

                <input
                    type='password'
                    placeholder='Enter your password'
                    name='password'
                    value={password}
                    className='auth__input'
                    onChange={handleInputChange}
                />


                <button className='btn btn-primary btn-block' disabled={loading} type='submit'>
                    Login
                </button>

            </form>
            <div className='auth__social-networks animate__animated animate__fadeIn animate__fast'>
                <h2>Login with social networks</h2>
                <div className='google-btn' onClick={handleGoogleLogin}>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>
            </div>
            <Link className='link' to='/auth/register'>
                Create new account
            </Link>
        </>
    )
}
