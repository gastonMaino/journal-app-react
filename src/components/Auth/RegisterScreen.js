import React from 'react';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmail } from '../../actions/auth';
// import { startLoginEmailPassword } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { msgError } = useSelector(state => state.ui);


    const [values, handleInputChange] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    })

    const { name, email, password, password2 } = values;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmail(name, password, email));
        }
    }

    const isFormValid = () => {

        if (name.length === 0) {
            dispatch(setError('name is required'));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError('email is no valid'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            dispatch(setError('password should be at 6 characters and match each other'));
            return false
        }

        dispatch(removeError());
        return true;
    }

    return (
        <>
            <h1 className='auth__title mb-4'>Register</h1>

            <form
                onSubmit={handleRegister}
                className='animate__animated animate__fadeIn animate__fast'
            >
                {
                    msgError
                    &&
                    <div className='auth__alert-error'>{msgError}</div>
                }
                <input
                    type='text'
                    placeholder='Name'
                    name='name'
                    value={name}
                    className='auth__input mb-3'
                    onChange={handleInputChange}
                />

                <input
                    type='text'
                    placeholder='Email'
                    name='email'
                    value={email}
                    className='auth__input mb-3'
                    onChange={handleInputChange}
                />

                <input
                    type='password'
                    placeholder='Password'
                    name='password'
                    value={password}
                    className='auth__input mb-3'
                    onChange={handleInputChange}
                />
                <input
                    type='password'
                    placeholder='Confirm password'
                    name='password2'
                    value={password2}
                    className='auth__input mb-3'
                    onChange={handleInputChange}
                />

                <button className='btn btn-primary btn-block' type='submit'>
                    Register
                </button>

            </form>

            <Link className='link mt-2' to='/auth/login'>
                Already register ?
            </Link>
        </>
    )
}
