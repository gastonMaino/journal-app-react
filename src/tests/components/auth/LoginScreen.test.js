import { mount } from "enzyme/build"
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom'

import { LoginScreen } from '../../../components/Auth/LoginScreen';
import { startGoogleLogin, startLoginEmailPassword } from "../../../actions/auth";

jest.mock('../../../actions/auth', () =>({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
);

describe('test in LoginScreen', () => {
    beforeEach(() => {
        store = mockStore(initState);
        jest.clearAllMocks();
    })

    test('should match to snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should shot action startGoogleLogin', () => {
        wrapper.find('.google-btn').prop('onClick')();

        expect(startGoogleLogin).toHaveBeenCalled();
    })

    test('should call with email and password ', () => {
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(startLoginEmailPassword).toHaveBeenLastCalledWith('', '');
    })
    
    

})
