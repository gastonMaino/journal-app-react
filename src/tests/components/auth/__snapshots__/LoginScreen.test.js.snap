// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in LoginScreen should match to snapshot 1`] = `
Array [
  <h1
    className="auth__title mb-4"
  >
    Login
  </h1>,
  <form
    className="animate__animated animate__fadeIn animate__fast"
    onSubmit={[Function]}
  >
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Enter your email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Enter your password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
  </form>,
  <div
    className="auth__social-networks animate__animated animate__fadeIn animate__fast"
  >
    <h2>
      Login with social networks
    </h2>
    <div
      className="google-btn"
      onClick={[Function]}
    >
      <div
        className="google-icon-wrapper"
      >
        <img
          alt="google button"
          className="google-icon"
          src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
        />
      </div>
      <p
        className="btn-text"
      >
        <b>
          Sign in with google
        </b>
      </p>
    </div>
  </div>,
  <a
    className="link"
    href="/auth/register"
    onClick={[Function]}
  >
    Create new account
  </a>,
]
`;
